plugins/processors/dcos_metadata/dcos_metadata.go

```diff
@@ -0,0 +1,273 @@
+package dcos_metadata
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"sync"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/internal"
+	"github.com/influxdata/telegraf/plugins/processors"
+
+	"github.com/mesos/mesos-go/api/v1/lib"
+	"github.com/mesos/mesos-go/api/v1/lib/agent"
+	"github.com/mesos/mesos-go/api/v1/lib/agent/calls"
+	"github.com/mesos/mesos-go/api/v1/lib/httpcli"
+	"github.com/mesos/mesos-go/api/v1/lib/httpcli/httpagent"
+)
+
+type DCOSMetadata struct {
+	MesosAgentUrl string
+	Timeout       internal.Duration
+	RateLimit     internal.Duration
+	containers    map[string]containerInfo
+	mu            sync.Mutex
+	once          Once
+}
+
+// containerInfo is a tuple of metadata which we use to map a container ID to
+// information about the task, executor and framework.
+type containerInfo struct {
+	containerID   string
+	taskName      string
+	executorName  string
+	frameworkName string
+	taskLabels    map[string]string
+}
+
+const sampleConfig = `
+## The URL of the local mesos agent
+mesos_agent_url = "http://$NODE_PRIVATE_IP:5051"
+## The period after which requests to mesos agent should time out
+timeout = "10s"
+## The minimum period between requests to the mesos agent
+rate_limit = "5s"
+`
+
+// SampleConfig returns the default configuration
+func (dm *DCOSMetadata) SampleConfig() string {
+	return sampleConfig
+}
+
+// Description returns a one-sentence description of dcos_metadata
+func (dm *DCOSMetadata) Description() string {
+	return "Plugin for adding metadata to dcos-specific metrics"
+}
+
+// Apply the filter to the given metrics
+func (dm *DCOSMetadata) Apply(in ...telegraf.Metric) []telegraf.Metric {
+	// stale tracks whether our container cache is stale
+	stale := false
+
+	for _, metric := range in {
+		// Ignore metrics without container_id tag
+		if cid, ok := metric.Tags()["container_id"]; ok {
+			if c, ok := dm.containers[cid]; ok {
+				// Data for this container was cached
+				for k, v := range c.taskLabels {
+					metric.AddTag(k, v)
+				}
+				metric.AddTag("service_name", c.frameworkName)
+				if c.executorName != "" {
+					metric.AddTag("executor_name", c.executorName)
+				}
+				metric.AddTag("task_name", c.taskName)
+			} else {
+				log.Printf("I! Information for container %q was not found in cache", cid)
+				stale = true
+			}
+		}
+	}
+
+	if stale {
+		go dm.refresh()
+	}
+
+	return in
+}
+
+func (dm *DCOSMetadata) refresh() {
+	// Subsequent calls to refresh() will be ignored until the RateLimit period
+	// has expired
+	go func() {
+		time.Sleep(dm.RateLimit.Duration)
+		dm.once.Reset()
+	}()
```

Comment:
Looks like this is intended to schedule a reset only when we execute the inlined function below. If that's the case, should it be moved inside that function? Here I think the reset will be scheduled even if we skip the refresh, which is going to change how rate limiting behaves.